src/workerd/io/actor-sqlite.c++

```diff
@@ -331,11 +452,52 @@ void ActorSqlite::shutdown(kj::Maybe<const kj::Exception&> maybeException) {
 }
 
 kj::Maybe<kj::Own<void>> ActorSqlite::armAlarmHandler(kj::Date scheduledTime, bool noCache) {
-  return hooks.armAlarmHandler(scheduledTime, noCache);
+  ensureAlarmTimeInitialized();
+
+  KJ_ASSERT(!currentAlarmTime.is<DeferredAlarmDelete>());
+
+  bool alarmDeleteNeeded = true;
+  KJ_IF_SOME(t, currentAlarmTime.tryGet<KnownAlarmTime>()) {
+    if (t.time != scheduledTime) {
+      if (t.status == KnownAlarmTime::Status::CLEAN) {
+        // If there's a clean scheduledTime that is different from ours, this run should be
+        // canceled.
+        return kj::none;
+      } else {
+        // There's a alarm write that hasn't been set yet pending for a time different than ours --
+        // We won't cancel the alarm because it hasn't been confirmed, but we shouldn't delete
+        // the pending write.
+        alarmDeleteNeeded = false;
+      }
+    }
+  }
+
+  // TODO(now): Here, similar to the ActorCache code, we're assuming that if the alarm time
+  // matches but is not in a clean state (i.e. is dirty or flushing), that it's OK to schedule the
+  // delete, and that it's OK to use the current alarm time to initialize the alarm state back to
+  // "clean" if the alarm is cancelled.  Is that OK?
+
+  if (alarmDeleteNeeded) {
+    currentAlarmTime = DeferredAlarmDelete{
+      .status = DeferredAlarmDelete::Status::WAITING,
+      .timeToDelete = scheduledTime,
+    };
+  }
+
+  static const DeferredAlarmDeleter disposer;
+  return kj::Own<void>(this, disposer);
 }
 
 void ActorSqlite::cancelDeferredAlarmDeletion() {
-  hooks.cancelDeferredAlarmDeletion();
+  if (currentAlarmTime.is<DeferredAlarmDelete>()) {
+    // TODO(now): Should we do anything special if client tries to cancel alarm deletion after
```

Comment:
That'd be an invalid use of the API, so hopefully we don't have to worry about it. But reporting an error to sentry if it happens might not be a bad idea.